### PR TITLE
WIP: Refactoring scene tree to make unordered set of unique pointers

### PR DESCRIPTION
**Pull request by aaryaag03**

### CODE REVIEW

**Review by r41k0u — CHANGES_REQUESTED** [review]
`editor.app.json` and `imgui.ini` must not be pushed

**Comment by r41k0u** [comment]
Please give a better PR name like `WIP: Refactoring scene tree to use unordered_set of unique_ptr` or something like that

File: rootex/framework/scene.cpp [hunk 1]
```diff
@@ -249,28 +249,27 @@ bool Scene::snatchChild(Scene* child)
 	}
 
 	Unordered_set<Ptr<Scene>>& children = child->getParent()->getChildren();
-	for (auto&& child_scene : children)
+	for (auto& child_scene : children)
```
Comment (Aviii06) [inline rootex/framework/scene.cpp:252]: is there a reason to use an lvalue reference instead of rvalue?

File: rootex/framework/scene.cpp [hunk 2]
```diff
@@ -307,9 +306,8 @@ bool Scene::addChild(Ptr<Scene>& child)
 		}
 		m_ChildrenScenes.insert(std::move(child));
 
-		auto it = m_ChildrenScenes.end();
-		--it;
-		ScriptSystem::GetSingleton()->addEnterScriptEntity(&(*it)->getEntity());
+		//aarya
```
Comment (Aviii06) [inline rootex/framework/scene.cpp:309]: ?

File: rootex/framework/scene.h [hunk 3]
```diff
@@ -1,98 +1,98 @@
...
-	ImportStyle getImportStyle() const { return m_ImportStyle; }
-	String getScenePath() const { return m_SceneFile; }
-	Scene* getParent() const { return m_ParentScene; }
-	Entity& getEntity() { return m_Entity; }
-	const String& getName() const { return m_Name; }
-	const String& getFullName() const { return m_FullName; }
-	void setFullName(String& name) { m_FullName = name; }
-	SceneSettings& getSettings() { return m_Settings; }
-};
+#pragma once
```
Comment (Aviii06) [inline rootex/framework/scene.h:1]: fix this, your diff shouldn't be the entire file.

File: rootex/script/interpreter.cpp [hunk 4]
```diff
@@ -124,68 +124,44 @@ void LuaInterpreter::registerTypes()
...
-		    { return l * r; },
-		    sol::meta_function::division, [](Vector2& l, float r)
-		    { return l / r; },
-		    sol::meta_function::division, [](Vector2& l, Vector2& r)
-		    { return l / r; });
-		vector2["dot"] = [](const Vector2& l, const Vector2& r)
-		{ return l.Dot(r); };
-		vector2["cross"] = [](const Vector2& l, const Vector2& r)
-		{ return l.Cross(r); };
+		    sol::meta_function::addition, [](Vector2& l, Vector2& r) { return l + r; },
```
Comment (Aviii06) [inline rootex/script/interpreter.cpp:127]: Run a linter here to have this being the original, ideally your diff should only contain things which you've changed functionally or semantiaclly.

File: rootex/framework/scene.cpp (hunk 1 above)
Comment (twaritwaikar) [reply rootex/framework/scene.cpp:252]: This is the universal binder, not an rvalue reference. However, this line diff doesn't seem to change anything.

File: rootex/common/types.h [hunk 5]
```diff
@@ -96,7 +96,12 @@ using Optional = std::optional<T>;
 #include <vector>
 /// std::vector
 template <class T>
-using Vector = std::vector<T>;
+using Vector = std::vector<T>;
+
+#include <unordered_set>
+/// std::unordered_set
+template <class T>
+using Unordered_set = std::unordered_set<T>;
```
Comment (twaritwaikar) [inline rootex/common/types.h:104]: Does this code even compile? `std::unordered_set` requires 2 template arguments.

File: rootex/framework/scene.h [hunk 6]
```diff
@@ -51,7 +51,7 @@ class Scene
 	Entity m_Entity;
 
 	Scene* m_ParentScene = nullptr;
-	Vector<Ptr<Scene>> m_ChildrenScenes;
+	Unordered_set<Ptr<Scene>> m_ChildrenScenes;
```
Comment (twaritwaikar) [inline rootex/framework/scene.h:54]: Why use an unordered set? What are the advantages of using one here?

Comment (Bashar-Ahmed) [reply rootex/framework/scene.h:54]: Basically, we needed O(1) addressing along with O(1) deletion. Based on this, the most ideal data structure seemed to be an unodered_set. O(1) deletion was needed because currently in many functions we loop over the complete vector which isn't necessary.

File: rootex/common/types.h (hunk 5 above)
Comment (Bashar-Ahmed) [reply rootex/common/types.h:104]: Don't we only need to provide the Key type and the rest (hash, key_equal, allocator) are picked as the default ones?
ref: https://en.cppreference.com/w/cpp/container/unordered_set

File: rootex/framework/scene.h (hunk 6 above)
Comment (twaritwaikar) [reply rootex/framework/scene.h:54]: @Bashar-Ahmed Is there any performance testing done to see if using a cache inefficient data structure but with a theoretically better complexity is performing better than a vector?

Also, do you want to save the order in which the children are added in a parent?

File: rootex/common/types.h (hunk 5 above)
Comment (twaritwaikar) [reply rootex/common/types.h:104]: My bad, confused this one with an unordered_map. You can resolve this conversation

File: rootex/framework/scene.h (hunk 6 above)
Comment (Bashar-Ahmed) [reply rootex/framework/scene.h:54]: No, as of now no performance testing has been done. Will have to look into it and get back to you. Also caching due to vector would've been minimal, if any, due to the fact that it was a vector of ptr, instead of the actual scenes. So when fetching the address,caching would be useful but when getting the object the caching wouldn't be effective. (Had a discussion regarding this with @sin3point14 ). Apologies if this approach is flawed, and do correct me. 

As far as the order is concerned, we looked at this aspect and weren't able to find any viable usecase where saving the order of scenes would be necessary

Comment (twaritwaikar) [reply rootex/framework/scene.h:54]: One way the order is important is when you need to iterate over the children to display them on the editor. Seeing the order of the children change as you make any changes in the scene hierarchy can get confusing.

**Comment by sin3point14** [comment]
I talked about a probably approach with archit which would need minimal
changes and should remove those pesky edge cases.
Only allow saving the root scene’s children and don’t allow 2 scenes of the
same path to be attached to the root. Ofc if you must attach 2 scenes of
the same path, you still can 2 make 2 empty scene children of the root and
add those 2 scenes as their children.
Since we serialise(afaik) all the children into the file of the direct root
children, saved scenes should only be treated as “templates” ie you can
instantiate them as child of some scene and then change the properties.
Saving the root child parent node(the node in the parent hierarchy which is
also a root child) will result in the scene of that root child to contain
all the data of this instantiated child scene. Another way to think about
this is that changing the contents of that child scene file would have no
effect on this attached instance, it lost all connections to its source
file and the moment it was attached as a non root child.

tldr: only allow saving of direct root child scenes ie root child scenes
are the only scenes who know their original file path and disallow multiple
root children from same path. This goes well with rootex philosophy of
everything can be uniquely identified with its file path.
